text/0042-partial-fill-rules.md

```diff
@@ -0,0 +1,296 @@
+///   `rB`.
+fn get_blinding_factors(amount_shared_secret: &[u8; 32]) -> (u64, u64, Scalar) {
+    // Use HKDF-SHA512 to produce blinding factors for value, token id, and
+    // commitment
+    let kdf = Hkdf::<Sha512>::new(
+        Some(AMOUNT_BLINDING_FACTORS_DOMAIN_TAG),
+        amount_shared_secret,
+    );
+
+    let mut value_mask = [0u8; 8];
+    kdf.expand(AMOUNT_VALUE_DOMAIN_TAG.as_bytes(), &mut value_mask)
+        .expect("Digest output size is insufficient");
+
+    let mut token_id_mask = [0u8; 8];
+    kdf.expand(AMOUNT_TOKEN_ID_DOMAIN_TAG.as_bytes(), &mut token_id_mask)
+        .expect("Digest output size is insufficient");
+
+    let mut scalar_blinding_bytes = [0u8; 64];
+    kdf.expand(
+        AMOUNT_BLINDING_DOMAIN_TAG.as_bytes(),
+        &mut scalar_blinding_bytes,
+    )
+    .expect("Digest output size is insufficient");
+
+    (
+        u64::from_le_bytes(value_mask),
+        u64::from_le_bytes(token_id_mask),
+        Scalar::from_bytes_mod_order_wide(&scalar_blinding_bytes),
+    )
+}
+```
+
+At the level of protobuf, the `TxOut.masked_amount` field is replaced with a `oneof`, selecting between V1 and V2.
+`MaskedAmountV1` uses the historical derivations and must be supported by all clients indefinitely, as outputs from block versions <=2 always use this derivation.
+`MaskedAmountV2` uses the new derivation and is mandatory for all `TxOut`'s from block version 3 and on.
+
+## Partial fill rules schema and semantics
+
+
+The proposed changes to the `TxIn` structure and the input rules are captured in this diagram:
+
+```mermaid
+classDiagram
+direction BT
+InputRules --> TxIn
+RevealedTxOut --> InputRules
+
+TxIn: Vec~TxOut~ ring
+TxIn: Vec~...~ membership_proofs
+TxIn: Option~InputRules~ input_rules
+
+InputRules: Vec~TxOut~ required_outputs
+InputRules: uint64 max_tombstone_block
+InputRules: Vec~RevealedTxOut~ partial_fill_outputs *new*
+InputRules: RevealedTxOut partial_fill_change *new*
+InputRules: uint64 min_partial_fill_value *new*
+
+RevealedTxOut: TxOut tx_out
+RevealedTxOut: bytes amount_shared_secret
+```
+
+Recall that the `InputRules` are created (and signed) by the **originator**, and this TxIn is added to a transaction
+by the **counterparty**, who is also obligated to add TxOut's to the TxPrefix which match to the partial fill change
+and partial fill outputs appropriately. (This is actually done under the hood by the transaction builder.)
+
+The consensus enclave must enforce any input rules, including these new partial fill rules. This is done in the following way.
+
+1. If `partial_fill_change` is not present, then no other partial fill rules may be used, and any other partial fill fields must be defaulted.
+1. If `partial_fill_change` is present, then decrypting the amount and token id using `amount_shared_secret` must succeed. The partial fill change
+   may not have a 0 value (to help clients avoid division by zero).
```

Comment:
You're right, we have this check in the input rules code:

https://github.com/mobilecoinfoundation/mobilecoin/blob/117e2c27900e9b22514d5eb69401391369cf1240/transaction/core/src/input_rules.rs#L104

I guess what's interesting here is, the enclave doesn't actually do any division, since it multiplies things out. So the enclave doesn't actually care itself right now if the partial fill change is zero.

I think what we should do is:

* Insert a check for the partial fill change being zero into the input rules verification. Even if it's not really required to protect the enclave, it might be simpler to explain the spec if we just enforce this everywhere.
* At line 104 we should insert a code comment that this prevents a later subtraction expression from panicking.

Reply:
Checking in the enclave for consistency sounds good.

Just to double check, is the panic at [138](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/transaction/core/src/input_rules.rs#L138) the only one that prevents creating mob or is there still a subsequent check with the RingCT logic (that I still need to find time to get a high level understanding of)

Reply:
The RingCT logic ensures that every transaction is "balanced" -- the value of every output in tx prefix + the fee output = the value of every true input. That check happens in `transaction/core/src/ring_ct/signature_rct_bulletproofs.rs` in the verify function somewhere